candle-metal-kernels/src/lib.rs

```diff
@@ -1567,7 +1522,176 @@ pub fn call_upsample_nearest_2d(
     encoder.use_resource(input, metal::MTLResourceUsage::Read);
     encoder.use_resource(output, metal::MTLResourceUsage::Write);
     encoder.dispatch_thread_groups(thread_group_count, thread_group_size);
-    encoder.update_fence(&kernels.fence);
+    encoder.end_encoding();
+
+    Ok(())
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum GgmlDType {
+    Q4_0,
```

Comment:
https://github.com/rustformers/llm use an FFI here to keep in line, something to consider for the future.

Reply:
Why would we need FFI ? This is purely for internal use to call the correct kernel with the correct warps. (Each kernel is written with specific warps in mind)